lightning/common/security.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/http/httptest"
+
+	"github.com/go-sql-driver/mysql"
+	"github.com/pingcap/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+// TLS
+type TLS struct {
+	inner  *tls.Config
+	client *http.Client
+	url    string
+}
+
+// NewTLS constructs a new HTTP client with TLS configured with the CA,
+// certificate and key paths.
+//
+// If the CA path is empty, returns an instance where TLS is disabled.
+func NewTLS(caPath, certPath, keyPath, host string) (*TLS, error) {
+	if len(caPath) == 0 {
+		return &TLS{
+			inner:  nil,
+			client: &http.Client{},
+			url:    "http://" + host,
+		}, nil
+	}
+
+	// Load the client certificates from disk
+	var certificates []tls.Certificate
+	if len(certPath) != 0 && len(keyPath) != 0 {
+		cert, err := tls.LoadX509KeyPair(certPath, keyPath)
+		if err != nil {
+			return nil, errors.Annotate(err, "could not load client key pair")
+		}
+		certificates = []tls.Certificate{cert}
+	}
+
+	// Create a certificate pool from CA
+	certPool := x509.NewCertPool()
+	ca, err := ioutil.ReadFile(caPath)
+	if err != nil {
+		return nil, errors.Annotate(err, "could not read ca certificate")
+	}
+
+	// Append the certificates from the CA
+	if !certPool.AppendCertsFromPEM(ca) {
+		return nil, errors.New("failed to append ca certs")
+	}
+
+	inner := &tls.Config{
+		Certificates: certificates,
+		RootCAs:      certPool,
+		NextProtos:   []string{"h2", "http/1.1"},
```

Comment:
what's the usage of this

Reply:
to allow HTTP/2.

Reply:
https://pkg.go.dev/net/http?tab=doc#Serve

 HTTP/2 support is only enabled if the Listener returns *tls.Conn connections and they were configured with "h2" in the TLS Config.NextProtos.